pkg/gadget-collection/gadgets/top/ebpf/gadget.go

```diff
@@ -0,0 +1,339 @@
+	intervalSeconds := types.IntervalDefault
+	sortBy := types.SortByDefault
+
+	if trace.Spec.Parameters != nil {
+		params := trace.Spec.Parameters
+		var err error
+
+		if val, ok := params[types.MaxRowsParam]; ok {
+			maxRows, err = strconv.Atoi(val)
+			if err != nil {
+				trace.Status.OperationError = fmt.Sprintf("%q is not valid for %s: %v", val, types.MaxRowsParam, err)
+				return
+			}
+		}
+
+		if val, ok := params[types.IntervalParam]; ok {
+			intervalSeconds, err = strconv.Atoi(val)
+			if err != nil {
+				trace.Status.OperationError = fmt.Sprintf("%q is not valid for %s: %v", val, types.IntervalParam, err)
+				return
+			}
+		}
+
+		if val, ok := params[types.SortByParam]; ok {
+			sortBy, err = types.ParseSortBy(val)
+			if err != nil {
+				trace.Status.OperationError = fmt.Sprintf("%q is not valid for %s: %v", val, types.SortByParam, err)
+				return
+			}
+		}
+	}
+
+	t.maxRows = maxRows
+	t.intervalSeconds = intervalSeconds
+	t.sortBy = sortBy
+
+	// Enable stats collection
+	err := bpfstats.EnableBPFStats()
+	if err != nil {
+		trace.Status.OperationError = err.Error()
+		return
+	}
+
+	t.useFallbackIterator = false
+
+	// To resolve pids, we will first try to iterate using a bpf
+	// program. If that doesn't work, we will fall back to scanning
+	// all used fds in all processes /proc/$pid/fdinfo/$fd.
+	iter, err := piditer.NewTracer()
+	if err != nil {
+		t.useFallbackIterator = true
+	} else {
+		t.iter = iter
+	}
+
+	t.started = true
+
+	go t.run()
+
+	trace.Status.State = "Started"
+}
+
+func (t *Trace) Stop(trace *gadgetv1alpha1.Trace) {
+	if !t.started {
+		trace.Status.OperationError = "Not started"
+		return
+	}
+
+	err := t.stop()
+	if err != nil {
+		trace.Status.OperationWarning = err.Error()
+	}
+
+	trace.Status.State = "Stopped"
+}
+
+func (t *Trace) stop() error {
+	t.donechan <- true
+	t.started = false
+	return bpfstats.DisableBPFStats()
+}
+
+func (t *Trace) run() {
+	if !t.useFallbackIterator {
+		defer t.iter.Close()
+	}
+
+	timer := time.NewTicker(time.Second * time.Duration(t.intervalSeconds))
+	for {
+		select {
+		case <-t.donechan:
+			return
+		case <-timer.C:
+			stats := make([]types.Stats, 0)
+
+			var err error
+			var prog *ebpf.Program
+			var pids []*piditer.PidIterEntry
+			curID := ebpf.ProgramID(0)
+			nextID := ebpf.ProgramID(0)
+
+			for {
+				nextID, err = ebpf.ProgramGetNextID(curID)
+				if errors.Is(err, os.ErrNotExist) {
+					break
+				}
+				if nextID <= curID {
+					break
+				}
+				curID = nextID
+				prog, err = ebpf.NewProgramFromID(curID)
+				if err != nil {
+					continue
+				}
+				pi, err := prog.Info()
+				if err != nil {
+					prog.Close()
+					continue
+				}
+				rt, _ := pi.Runtime()
+				rc, _ := pi.RunCount()
+
+				stats = append(stats, types.Stats{
+					ProgramID: uint32(curID),
+					Pids:      nil,
+					Name:      pi.Name,
+					Type:      pi.Type.String(),
+					Runtime:   int64(rt),
+					RunCount:  rc,
+				})
+
+				prog.Close()
+			}
```

Comment:
Yeah, totally agree. Especially since introducing the fallback, this has grown. I have extracted the pid aggregation into separate functions - should be more readable now.